Index: mainwnd.c
===================================================================
--- mainwnd.c	(revision 3671)
+++ mainwnd.c	(working copy)
@@ -126,6 +126,11 @@
     __in_opt PVOID Context
     );
 
+BOOLEAN PhpSignedMsProcessTreeFilter(
+    __in PPH_PROCESS_NODE ProcessNode,
+    __in_opt PVOID Context
+    );
+
 PPH_PROCESS_ITEM PhpGetSelectedProcess();
 
 VOID PhpGetSelectedProcesses(
@@ -259,6 +264,7 @@
 static ULONG SelectedUserSessionId;
 
 static PPH_PROCESS_TREE_FILTER_ENTRY CurrentUserFilterEntry = NULL;
+static PPH_PROCESS_TREE_FILTER_ENTRY SignedMsFilterEntry = NULL;
 
 BOOLEAN PhMainWndInitialization(
     __in INT ShowCommand
@@ -708,6 +714,37 @@
                         );
                 }
                 break;
+            case ID_VIEW_HIDESYSTEMPROCESSES:
+                {
+                    if (!SignedMsFilterEntry)
+                    {
+                        if (!PhEnableProcessQueryStage2)
+                        {
+                            PhShowInformation(
+                                hWnd,
+                                L"Processes cannot be verified to be actual system processes because "
+                                L"digital signature checking is disabled. "
+                                L"You can enable it in Options > Advanced."
+                                );
+                        }
+
+                        SignedMsFilterEntry = PhAddProcessTreeFilter(PhpSignedMsProcessTreeFilter, NULL);
+                    }
+                    else
+                    {
+                        PhRemoveProcessTreeFilter(SignedMsFilterEntry);
+                        SignedMsFilterEntry = NULL;
+                    }
+
+                    PhApplyProcessTreeFilters();
+
+                    CheckMenuItem(
+                        PhMainWndMenuHandle,
+                        ID_VIEW_HIDESYSTEMPROCESSES,
+                        SignedMsFilterEntry ? MF_CHECKED : MF_UNCHECKED
+                        );
+                }
+                break;
             case ID_VIEW_ALWAYSONTOP:
                 {
                     BOOLEAN topMost;
@@ -2732,6 +2769,44 @@
     return TRUE;
 }
 
+BOOLEAN PhpSignedMsProcessTreeFilter(
+    __in PPH_PROCESS_NODE ProcessNode,
+    __in_opt PVOID Context
+    )
+{
+    BOOLEAN isRedmond;
+
+    isRedmond = FALSE;
+
+    if (!ProcessNode->ProcessItem->FileName)
+        return TRUE;
+
+    if (PhEnableProcessQueryStage2)
+    {
+        if (ProcessNode->ProcessItem->VerifyResult != VrTrusted)
+            return TRUE;
+        if (!ProcessNode->ProcessItem->VerifySignerName)
+            return TRUE;
+
+        if (PhEqualString2(ProcessNode->ProcessItem->VerifySignerName, L"Microsoft Windows", FALSE))
+            isRedmond = TRUE;
+    }
+    else
+    {
+        // No signatures, so check the version info.
+        if (
+            ProcessNode->ProcessItem->VersionInfo.CompanyName &&
+            PhEqualString2(ProcessNode->ProcessItem->VersionInfo.CompanyName, L"Microsoft Corporation", FALSE)
+            )
+            isRedmond = TRUE;
+    }
+
+    if (isRedmond && PhStartsWithString2(ProcessNode->ProcessItem->FileName, USER_SHARED_DATA->NtSystemRoot, TRUE))
+        return FALSE;
+
+    return TRUE;
+}
+
 PPH_PROCESS_ITEM PhpGetSelectedProcess()
 {
     return PhGetSelectedProcessItem();
@@ -3956,6 +4031,9 @@
     )
 {
     PhUpdateProcessNode(PhFindProcessNode(ProcessItem->ProcessId));
+
+    if (SignedMsFilterEntry)
+        PhApplyProcessTreeFilters();
 }
 
 VOID PhMainWndOnProcessRemoved(
Index: ProcessHacker.rc
===================================================================
--- ProcessHacker.rc	(revision 3671)
+++ ProcessHacker.rc	(working copy)
@@ -102,6 +102,7 @@
             MENUITEM "Physical Memory History",     ID_TRAYICONS_PHYSICALMEMORYHISTORY
         END
         MENUITEM "Hide Processes From Other Users", ID_VIEW_HIDEPROCESSESFROMOTHERUSERS
+        MENUITEM "Hide System Processes",       ID_VIEW_HIDESYSTEMPROCESSES
         MENUITEM SEPARATOR
         MENUITEM "Always on Top",               ID_VIEW_ALWAYSONTOP
         POPUP "Opacity"
Index: resource.h
===================================================================
--- resource.h	(revision 3671)
+++ resource.h	(working copy)
@@ -610,6 +610,7 @@
 #define ID_HACKER_PLUGINS               40231
 #define ID_VIEW_HIDEPROCESSESFROMOTHERUSERS 40232
 #define ID_THREAD_AFFINITY              40233
+#define ID_VIEW_HIDESYSTEMPROCESSES     40234
 #define IDDYNAMIC                       50000
 #define IDPLUGINS                       55000
 #define IDPLUGINS_END                   56000
@@ -619,7 +620,7 @@
 #ifdef APSTUDIO_INVOKED
 #ifndef APSTUDIO_READONLY_SYMBOLS
 #define _APS_NEXT_RESOURCE_VALUE        191
-#define _APS_NEXT_COMMAND_VALUE         40234
+#define _APS_NEXT_COMMAND_VALUE         40235
 #define _APS_NEXT_CONTROL_VALUE         1271
 #define _APS_NEXT_SYMED_VALUE           132
 #endif
